clients/da-vinci-client/src/main/java/com/linkedin/davinci/notifier/DaVinciPushStatusUpdateTask.java

```diff
@@ -0,0 +1,115 @@
+
+
+/**
+ * This is a scheduler for sending batching push status in DaVinci.
+ */
+public class DaVinciPushStatusUpdateTask {
+  private final Version version;
+  private final PushStatusStoreWriter pushStatusStoreWriter;
+  private boolean batchPushStartSignalSent;
+  private boolean batchPushEndSignalSent;
+  private final long daVinciPushStatusCheckIntervalInMs;
+  private final Supplier<Boolean> areAllPartitionFuturesCompletedSuccessfully;
+  private final Map<Integer, ExecutionStatus> partitionStatus = new VeniceConcurrentHashMap<>();
+  // Executor for scheduling tasks
+  private final ScheduledExecutorService scheduler =
+      Executors.newScheduledThreadPool(1, new DaemonThreadFactory("davinci-push-status-update-task-scheduler"));
+
+  public DaVinciPushStatusUpdateTask(
+      Version version,
+      long daVinciPushStatusCheckIntervalInMs,
+      PushStatusStoreWriter pushStatusStoreWriter,
+      Supplier<Boolean> areAllPartitionFuturesCompletedSuccessfully) {
+    this.version = version;
+    this.batchPushStartSignalSent = false;
+    this.batchPushEndSignalSent = false;
+    this.daVinciPushStatusCheckIntervalInMs = daVinciPushStatusCheckIntervalInMs;
+    this.pushStatusStoreWriter = pushStatusStoreWriter;
+    this.areAllPartitionFuturesCompletedSuccessfully = areAllPartitionFuturesCompletedSuccessfully;
+  }
+
+  public boolean isBatchPushStartSignalSent() {
+    return batchPushStartSignalSent;
+  }
+
+  public boolean isBatchPushEndSignalSent() {
+    return batchPushEndSignalSent;
+  }
+
+  public void batchPushStartSignalSent() {
+    this.batchPushStartSignalSent = true;
+  }
+
+  public void batchPushEndSignalSent() {
+    this.batchPushEndSignalSent = true;
+  }
+
+  public void updatePartitionStatus(int partition, ExecutionStatus status) {
+    partitionStatus.put(partition, status);
+  }
+
+  private void maybeSendBatchingStatus() {
+    // First, check if the batch push start signal has been sent, if not, send out STARTED status for this version
+    // immediately, regardless if there is any partition subscription happened (this is also needed for delayed version
+    // swap for DaVinci during target colo pushes)
+    // If STARTED is sent, and if COMPLETED is not sent, check if all partitions are in COMPLETED status. If so, send
+    // the COMPLETED status.
+    if (!isBatchPushStartSignalSent()) {
+      pushStatusStoreWriter.writeVersionLevelPushStatus(
+          version.getStoreName(),
+          version.getNumber(),
+          ExecutionStatus.STARTED,
+          getTrackedPartitions());
+      batchPushStartSignalSent();
+    } else if (!isBatchPushEndSignalSent() && areAllPartitionsOnSameTerminalStatus(ExecutionStatus.COMPLETED)
+        && areAllPartitionFuturesCompletedSuccessfully.get()) {
+      pushStatusStoreWriter.writeVersionLevelPushStatus(
+          version.getStoreName(),
+          version.getNumber(),
+          ExecutionStatus.COMPLETED,
+          getTrackedPartitions());
```

Comment:
What if some DaVinci partition ingestion failed, so this update task will never send `ERROR` status, which is also a terminal status?
Maybe we should check whether all partition status hit terminal status or not, which can be `COMPLETED` or `ERROR`, and then decide the aggregated status and then send it to the push status system store.

Reply:
Good catch! Added failure path handling.